### PR TITLE
WFLY-7115 KeyAffinityService blocks Infinispan's topology change thread

### PR DESCRIPTION
**Pull request by pferraro**
https://issues.redhat.com/browse/WFLY-7115

Replaces Infinispan's default KeyAffinityService with a custom implementation without the problem described in WFLY-7715.
Also encapsulates all key distribution logic into a single object to ensure agreement between disparate parties.

### CODE REVIEW

**Comment by pferraro** [comment]
@rhusar @rachmatowicz  Please review.

File: clustering/infinispan/spi/src/main/java/org/wildfly/clustering/infinispan/spi/affinity/DefaultKeyAffinityService.java [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+                return keys.poll(nanos, TimeUnit.NANOSECONDS);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
+        return null;
+    }
+
+    @TopologyChanged
+    public CompletionStage<Void> viewChanged(TopologyChangedEvent<?, ?> event) {
```
Comment (rhusar) [inline clustering/infinispan/spi/src/main/java/org/wildfly/clustering/infinispan/spi/affinity/DefaultKeyAffinityService.java:217]: This method should rather be called `topologyChanged(...)`.

**Review by rhusar — APPROVED** [review]
I think this looks good!

File: clustering/infinispan/spi/src/main/java/org/wildfly/clustering/infinispan/spi/affinity/DefaultKeyAffinityService.java (hunk 1 above)
Comment (pferraro) [reply clustering/infinispan/spi/src/main/java/org/wildfly/clustering/infinispan/spi/affinity/DefaultKeyAffinityService.java:217]: The name is arbitrary, but I agree with you.

**Review by rhusar — APPROVED** [review]
LGTM 👍

**Pull request merged**